packet/src/pacer.rs

```diff
@@ -0,0 +1,1032 @@
+
+/// The state of a single upstream queue.
+/// The pacer manages packets across several upstream queues.
+#[derive(Debug)]
+pub struct QueueState {
+    /// The unique identifier for this queue within the session.
+    pub id: QueueId,
+    /// The total size of queued payloads within the queue.
+    pub size: DataSize,
+    /// The total number of packets in the queue.
+    /// NB: This is not a [`usize`] because it will later be used to divide a [`Duration`], for which
+    /// [`usize`] isn't implement. If the queues end up with 2^32 packets something has gone very wrong
+    /// in any case.
+    pub packet_count: u32,
+    /// The total time the packets in the queue have spent queued.
+    pub total_queue_time: Duration,
+    /// The kind of packets this queue contains.
+    pub queue_kind: PacketKind,
+    // The last time a packet was sent from this queue.
+    last_send_time: Option<Instant>,
+    // The time the packet at the front of the queue was queued.
+    leading_packet_queue_time: Option<Instant>,
+}
+
+/// The outcome of a call to [`Pacer::poll_action`].
+#[derive(Debug, Clone, Copy)]
+pub enum PollOutcome {
+    /// The caller **MUST** poll the next packet from the queue that produced the contained [`QueueState`]
+    /// and send it.
+    PollQueue(QueueId),
+    /// The caller MUST do nothing for now.
+    Nothing,
+    // TODO: Produce padding when we are sending very infrequently or if we need it to probe.
+    // PollPadding(QS)
+}
+
+/// A null pacer that doesn't pace.
+#[derive(Debug, Default)]
+pub struct NullPacer {
+    last_sends: HashMap<QueueId, Instant>,
+    queue_states: Vec<QueueState>,
+    need_immediate_timeout: bool,
+}
+
+impl Pacer for NullPacer {
+    fn set_target_bitrate(&mut self, _target_bitrate: Bitrate) {
+        // We don't care
+    }
+
+    fn poll_timeout(&self) -> Option<Instant> {
+        if self.need_immediate_timeout {
+            self.last_sends.values().min().copied()
+        } else {
+            None
+        }
+    }
+
+    fn handle_timeout(&mut self, _now: Instant, iter: impl Iterator<Item = QueueState>) {
+        self.need_immediate_timeout = false;
+        self.queue_states.clear();
+        self.queue_states.extend(iter);
+    }
+
+    fn poll_action(&mut self) -> PollOutcome {
+        let non_empty_queues = self.queue_states.iter().filter(|q| q.packet_count > 0);
+        // Pick a queue using round robin, prioritize the least recently sent on queue.
+        let to_send_on = non_empty_queues.min_by_key(|q| self.last_sends.get(&q.id));
+
+        self.need_immediate_timeout = true;
+        to_send_on.into()
+    }
+
+    fn register_send(&mut self, now: Instant, _packet_size: DataSize, from: QueueId) {
+        let e = self.last_sends.entry(from).or_insert(now);
+        *e = now;
+    }
+}
+
+/// A leaky bucket pacer that can overshoot the target bitrate when required.
+pub struct LeakyBucketPacer {
+    /// Target bitrate.
+    target_bitrate: Bitrate,
```

Comment:
Again, I would suggest not using the name "target" here, since it would easily be confused with the output of the congestion controller.  Better to call this "pacing rate" or "data rate" or maybe "unadjusted_rate".